stan/math/prim/meta/require_generics.hpp

```diff
@@ -1,976 +1,37 @@
-
-template <template <class...> class TypeCheck, class... Check>
-using require_any_eigen_st
-    = require_any_t<is_eigen_scalar_check<TypeCheck, Check>...>;
-
-template <template <class...> class TypeCheck, class... Check>
-using require_any_not_eigen_st
-    = require_any_not_t<is_eigen_scalar_check<TypeCheck, Check>...>;
-
-template <template <class...> class TypeCheck, class... Check>
-using require_all_eigen_st
-    = require_all_t<is_eigen_scalar_check<TypeCheck, Check>...>;
-
-template <template <class...> class TypeCheck, class... Check>
-using require_all_not_eigen_st
-    = require_all_not_t<is_eigen_scalar_check<TypeCheck, Check>...>;
-
-template <template <class...> class TypeCheck, class... Check>
-using require_eigen_vector_st
-    = require_t<is_eigen_vector_scalar_check<TypeCheck, Check...>>;
-
-template <template <class...> class TypeCheck, class... Check>
-using require_not_eigen_vector_st
-    = require_not_t<is_eigen_vector_scalar_check<TypeCheck, Check...>>;
-
-template <template <class...> class TypeCheck, class... Check>
-using require_any_eigen_vector_st
-    = require_any_t<is_eigen_vector_scalar_check<TypeCheck, Check>...>;
-
-template <template <class...> class TypeCheck, class... Check>
-using require_any_not_eigen_vector_st
-    = require_any_not_t<is_eigen_vector_scalar_check<TypeCheck, Check>...>;
-
-template <template <class...> class TypeCheck, class... Check>
-using require_all_eigen_vector_st
-    = require_all_t<is_eigen_vector_scalar_check<TypeCheck, Check>...>;
-
-template <template <class...> class TypeCheck, class... Check>
-using require_all_not_eigen_vector_st
-    = require_all_not_t<is_eigen_vector_scalar_check<TypeCheck, Check>...>;
-
-template <typename Row, typename Col>
-using require_eigen_row_and_col_t = require_t<
-    math::conjunction<is_eigen_row_vector<Row>, is_eigen_col_vector<Col>>>;
-
-template <typename Row, typename Col>
-using require_not_eigen_row_and_col_t = require_not_t<
-    math::conjunction<is_eigen_row_vector<Row>, is_eigen_col_vector<Col>>>;
+STAN_ADD_REQUIRE_BINARY(same, std::is_same, require_std);
+STAN_ADD_REQUIRE_BINARY_INNER(same, std::is_same, require_std);
+
+STAN_ADD_REQUIRE_BINARY(convertible, std::is_convertible, require_std);
+STAN_ADD_REQUIRE_BINARY_INNER(convertible, std::is_convertible, require_std);
+
+STAN_ADD_REQUIRE_UNARY(arithmetic, std::is_arithmetic,
+                       require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY_INNER(arithmetic, std::is_arithmetic,
+                             require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY(floating_point, std::is_floating_point,
+                       require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY_INNER(floating_point, std::is_floating_point,
+                             require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY(index, std::is_integral, require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY_INNER(index, std::is_integral, require_stan_scalar_real);
```

Comment:
I know it is not part of this PR, but why do we have `require_index` for `is_integral`?

Reply:
tbh I put that in there whenever I was working on stan upstream stuff, I can change this